model/project_parser.go

```diff
@@ -1051,3 +1075,218 @@ func evaluateDependsOn(tse *tagSelectorEvaluator, tgse *tagSelectorEvaluator, vs
+	taskNameExist := map[string]bool{}
+	for _, task := range pp.Tasks {
+		taskNameExist[task.Name] = true
+	}
+	for _, task := range toMerge.Tasks {
+		if _, ok := taskNameExist[task.Name]; ok {
+			catcher.Errorf("task '%s' has been declared already", task.Name)
+		} else {
+			pp.Tasks = append(pp.Tasks, task)
+			taskNameExist[task.Name] = true
+		}
+	}
+
+	taskGroupNameExist := map[string]bool{}
+	for _, taskGroup := range pp.TaskGroups {
+		taskGroupNameExist[taskGroup.Name] = true
+	}
+	for _, taskGroup := range toMerge.TaskGroups {
+		if _, ok := taskGroupNameExist[taskGroup.Name]; ok {
+			catcher.Errorf("task group '%s' has been declared already", taskGroup.Name)
+		} else {
+			pp.TaskGroups = append(pp.TaskGroups, taskGroup)
+			taskGroupNameExist[taskGroup.Name] = true
+		}
+	}
+
+	parameterKeyExist := map[string]bool{}
+	for _, parameter := range pp.Parameters {
+		parameterKeyExist[parameter.Key] = true
+	}
+	for _, parameter := range toMerge.Parameters {
+		if _, ok := parameterKeyExist[parameter.Key]; ok {
+			catcher.Errorf("parameter key '%s' has been declared already", parameter.Key)
+		} else {
+			pp.Parameters = append(pp.Parameters, parameter)
+			parameterKeyExist[parameter.Key] = true
+		}
+	}
+
+	moduleExist := map[string]bool{}
+	for _, module := range pp.Modules {
+		moduleExist[module.Name] = true
+	}
+	for _, module := range toMerge.Modules {
+		if _, ok := moduleExist[module.Name]; ok {
+			catcher.Errorf("module '%s' has been declared already", module.Name)
+		} else {
+			pp.Modules = append(pp.Modules, module)
+			moduleExist[module.Name] = true
+		}
+	}
+
+	for key, val := range toMerge.Functions {
+		if _, ok := pp.Functions[key]; ok {
+			catcher.Errorf("function '%s' has been declared already", key)
+		} else {
+			pp.Functions[key] = val
+		}
+	}
+}
+
+func (pp *ParserProject) mergeUnordered(toMerge *ParserProject) {
+	pp.Ignore = append(pp.Ignore, toMerge.Ignore...)
+	pp.Loggers = mergeAllLogs(pp.Loggers, toMerge.Loggers)
+}
+
+func (pp *ParserProject) mergeOrderedUnique(toMerge *ParserProject, catcher grip.Catcher) {
+	if pp.Pre != nil && toMerge.Pre != nil {
+		catcher.New("pre can only be defined in one yaml")
+	} else if toMerge.Pre != nil {
+		pp.Pre = toMerge.Pre
+	}
+
+	if pp.Post != nil && toMerge.Post != nil {
+		catcher.New("post can only be defined in one yaml")
+	} else {
+		if toMerge.Post != nil {
+			pp.Post = toMerge.Post
+		}
+	}
+
+	if pp.Timeout != nil && toMerge.Timeout != nil {
+		catcher.New("timeout can only be defined in one yaml")
+	} else {
+		if toMerge.Timeout != nil {
+			pp.Timeout = toMerge.Timeout
+		}
+	}
+
+	if pp.EarlyTermination != nil && toMerge.EarlyTermination != nil {
+		catcher.New("early termination can only be defined in one yaml")
+	} else {
+		if toMerge.EarlyTermination != nil {
+			pp.EarlyTermination = toMerge.EarlyTermination
+		}
+	}
+}
+
+func (pp *ParserProject) mergeUnique(toMerge *ParserProject, catcher grip.Catcher) {
```

Comment:
rather than passing in a catcher, we should use a new catcher, and then just add it to the main one when we call the merge function

Reply:
each function now returns an error